File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/Nocks.java
package com.mobnetic.coinguardian.model.market;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.CurrencyPairInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.List;

public class Nocks extends Market {

	private final static String NAME = "Nocks";
	private final static String TTS_NAME = NAME;
	private final static String URL = "https://api.nocks.com/api/v2/trade-market/%1$s";
	private final static String URL_CURRENCY_PAIRS = "https://api.nocks.com/api/v2/trade-market";

	public Nocks() {
		super(NAME, TTS_NAME, null);
	}

	@Override
	public String getUrl(int requestId, CheckerInfo checkerInfo) {
		return String.format(URL, checkerInfo.getCurrencyPairId());
	}
	
	@Override
	protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker, CheckerInfo checkerInfo) throws Exception {
		final JSONObject dataObject = jsonObject.getJSONObject("data");

	    ticker.bid = getDoubleFromJsonObject(dataObject, "buy");
		ticker.ask = getDoubleFromJsonObject(dataObject, "sell");
		ticker.vol = getDoubleFromJsonObject(dataObject, "volume");
		ticker.high = getDoubleFromJsonObject(dataObject, "high");
		ticker.low = getDoubleFromJsonObject(dataObject, "low");
		ticker.last = getDoubleFromJsonObject(dataObject, "last");
	}

	private double getDoubleFromJsonObject(JSONObject jsonObject, String key) throws Exception {
	    return jsonObject.getJSONObject(key).getDouble("amount");
    }

    // ====================
    // Get currency pairs
    // ====================
    @Override
    public String getCurrencyPairsUrl(int requestId) {
        return URL_CURRENCY_PAIRS;
    }

    @Override
    protected void parseCurrencyPairsFromJsonObject(int requestId, JSONObject jsonObject, List<CurrencyPairInfo> pairs) throws Exception {
	    final JSONArray resultJsonArray = jsonObject.getJSONArray("data");

	    for(int i=0; i<resultJsonArray.length(); ++i) {
	        final JSONObject marketJsonObject = resultJsonArray.getJSONObject(i);
	        final String pairId = marketJsonObject.getString("code");
			String[] currencies = pairId.split("-");
			if(currencies.length >= 2) {
				pairs.add(new CurrencyPairInfo(
						currencies[0],
						currencies[1],
						pairId));
			}
        }
    }
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/Nocks.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.CurrencyPairInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import org.json.JSONObject

class Nocks : Market(NAME, TTS_NAME, null) {
    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return String.format(URL, checkerInfo.currencyPairId)
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        val dataObject = jsonObject.getJSONObject("data")
        ticker.bid = getDoubleFromJsonObject(dataObject, "buy")
        ticker.ask = getDoubleFromJsonObject(dataObject, "sell")
        ticker.vol = getDoubleFromJsonObject(dataObject, "volume")
        ticker.high = getDoubleFromJsonObject(dataObject, "high")
        ticker.low = getDoubleFromJsonObject(dataObject, "low")
        ticker.last = getDoubleFromJsonObject(dataObject, "last")
    }

    @Throws(Exception::class)
    private fun getDoubleFromJsonObject(jsonObject: JSONObject, key: String): Double {
        return jsonObject.getJSONObject(key).getDouble("amount")
    }

    // ====================
    // Get currency pairs
    // ====================
    override fun getCurrencyPairsUrl(requestId: Int): String? {
        return URL_CURRENCY_PAIRS
    }

    @Throws(Exception::class)
    override fun parseCurrencyPairsFromJsonObject(requestId: Int, jsonObject: JSONObject, pairs: MutableList<CurrencyPairInfo>) {
        val resultJsonArray = jsonObject.getJSONArray("data")
        for (i in 0 until resultJsonArray.length()) {
            val marketJsonObject = resultJsonArray.getJSONObject(i)
            val pairId = marketJsonObject.getString("code")
            val currencies = pairId.split("-".toRegex()).toTypedArray()
            if (currencies.size >= 2) {
                pairs.add(CurrencyPairInfo(
                        currencies[0],
                        currencies[1],
                        pairId))
            }
        }
    }

    companion object {
        private const val NAME = "Nocks"
        private const val TTS_NAME = NAME
        private const val URL = "https://api.nocks.com/api/v2/trade-market/%1\$s"
        private const val URL_CURRENCY_PAIRS = "https://api.nocks.com/api/v2/trade-market"
    }
}